### PR TITLE
Add support for S3 Server Access logs in AWS module

### PR DESCRIPTION
**Pull request by CarlosRS9**
|Related issue|
|---|
| #7811|

## Description

Hi team,

This PR adds support to `S3 Server Access` buckets as explained in #7811. The new bucket type is called `server_access` and works in the same way as the rest of the supported bucket types.

A new class called `AWSServiceAccess` which inherits from `AWSCustomBucket` has been added. This class contains custom `check_bucket` and `load_information_from_file` functions. The former was added to ensure the user could provide a preffix for their logs using the `path` tag while the latter is in charge of parsing and cleaning the AWS logs, as some fields may contain problematic characters such as empty spaces or `[`. The details are explained in the issue.

## Configuration example
```
<wodle name="aws-s3">
  <disabled>no</disabled>
  <interval>5m</interval>
  <bucket type="server_access">
    <name>wazuh-aws-wodle</name>
    <aws_profile>default</aws_profile>
    <path>access_logs</path>
    <only_logs_after>2020-JAN-01</only_logs_after>
  </bucket>
</wodle>
```
Note: the `path` tag is optional.

## Manual usage

In the same way as the rest of the buckets and services supported by our AWS module, this integration can be manually tested by using the following command:

```
/var/ossec/wodles/aws/aws-s3 -t server_access -b wazuh-aws-wodle-access-logs -a $ACCESS -k $KEY -d 2 -s 2021-APR-28
```

### CODE REVIEW

File: wodles/aws/aws_s3.py [hunk 1]
```diff
@@ -2138,6 +2138,159 @@ def load_information_from_file(self, log_key):
...
+    def load_information_from_file(self, log_key):
+        """Load data from a S3 access log file."""
+        def parse_line(line_):
+            def merge_values(delimiter='"', remove=False):
+                while next_ := next(it, None):
+                    value_list[-1] = f'{value_list[-1]} {next_}'
+                    try:
+                        if next_[-1] == delimiter:
+                            if remove:
+                                value_list[-1] = value_list[-1][1:-2]
```
Comment (Selutario) [inline wodles/aws/aws_s3.py:2257]: Shouldn't `value_list[-1][1:-2]` be this?: 
```suggestion
                                value_list[-1] = value_list[-1][1:-1]
```

For instance:

> value_list[-1]

```
'"PUT /wazuh-aws-wodle/kms_compress_encrypted/2020/10/28/10/firehose_kms-1-etc.zip HTTP/1.1"'
```

> value_list[-1][1:-2]
'PUT /wazuh-aws-wodle/kms_compress_encrypted/2020/10/28/10/firehose_kms-1-etc.zip HTTP/1.'

The last 1 of the string is being lost, not just the delimiter.

**Review by Selutario — CHANGES_REQUESTED** [review]
Consider adding some more comments so it is easier to understand what is the aim of functions like `merge_values()`

File: wodles/aws/aws_s3.py [hunk 2]
```diff
@@ -2138,6 +2138,159 @@ def load_information_from_file(self, log_key):
...
+
+    def __init__(self, **kwargs):
+        db_table_name = 's3_server_access'
+        AWSCustomBucket.__init__(self, db_table_name=db_table_name, **kwargs)
+
+    def iter_files_in_bucket(self, aws_account_id=None, aws_region=None):
+        def skip_if_old(downloaded_file):
+            date = re.search(r'(\d{4}-\d{2}-\d{2}).*', downloaded_file)
+            if date:
+                from datetime import datetime
```
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/aws/aws_s3.py:2151]: It was placed there during testing and forgotten. It will be moved

File: wodles/aws/aws_s3.py (hunk 1 above)
Comment (CarlosRS9) [reply wodles/aws/aws_s3.py]: That was introduced due to a bug that is now fixed. It is no longer needed. Your suggested change has been applied.

File: wodles/aws/aws_s3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarlosRS9) [reply wodles/aws/aws_s3.py]: removed

**Review by Selutario — APPROVED** [review]

File: wodles/aws/aws_s3.py [hunk 3]
```diff
@@ -2143,6 +2143,162 @@ def load_information_from_file(self, log_key):
...
+                debug("++ Found new log: {0}".format(bucket_file['Key']), 2)
+                # Get the log file from S3 and decompress it
+                log_json = self.get_log_file(aws_account_id, bucket_file['Key'])
+                self.iter_events(log_json, bucket_file['Key'], aws_account_id)
+                # Remove file from S3 Bucket
+                if self.delete_file:
+                    debug("+++ Remove file from S3 Bucket:{0}".format(bucket_file['Key']), 2)
+                    self.client.delete_object(Bucket=self.bucket, Key=bucket_file['Key'])
+                self.mark_complete(aws_account_id, aws_region, bucket_file)
+            # optimize DB
```
Comment (davidjiglesias) [inline wodles/aws/aws_s3.py:2190]:
```suggestion
            # Optimize DB
```

File: wodles/aws/aws_s3.py [hunk 4]
```diff
@@ -2143,6 +2143,162 @@ def load_information_from_file(self, log_key):
...
+            self.db_maintenance(aws_account_id=aws_account_id, aws_region=aws_region)
+            self.db_connector.commit()
+            # iterate if there are more logs
```
Comment (davidjiglesias) [inline wodles/aws/aws_s3.py:2193]:
```suggestion
            # Iterate if there are more logs
```

File: wodles/aws/aws_s3.py [hunk 5]
```diff
@@ -2143,6 +2143,162 @@ def load_information_from_file(self, log_key):
...
+                    debug("++ Found new log: {0}".format(bucket_file['Key']), 2)
+                    # Get the log file from S3 and decompress it
+                    log_json = self.get_log_file(aws_account_id, bucket_file['Key'])
+                    self.iter_events(log_json, bucket_file['Key'], aws_account_id)
+                    # Remove file from S3 Bucket
+                    if self.delete_file:
+                        debug("+++ Remove file from S3 Bucket:{0}".format(bucket_file['Key']), 2)
+                        self.client.delete_object(Bucket=self.bucket, Key=bucket_file['Key'])
+                    self.mark_complete(aws_account_id, aws_region, bucket_file)
+                # optimize DB
```
Comment (davidjiglesias) [inline wodles/aws/aws_s3.py:2226]:
```suggestion
                # Optimize DB
```

**Review by davidjiglesias — DISMISSED** [review]

**Review by Selutario — APPROVED** [review]

**Pull request merged**